packages/core/src/types.ts

```diff
@@ -178,6 +178,101 @@ export interface Profile {
   [claim: string]: unknown
 }
 
+export interface SignInCallbackParams<P = Profile, A = Account> {
+  user: User | AdapterUser
+  account: A | null
+  /**
+   * If OAuth provider is used, it contains the full
+   * OAuth profile returned by your provider.
+   */
+  profile?: P
+  /**
+   * If Email provider is used, on the first call, it contains a
+   * `verificationRequest: true` property to indicate it is being triggered in the verification request flow.
+   * When the callback is invoked after a user has clicked on a sign in link,
+   * this property will not be present. You can check for the `verificationRequest` property
+   * to avoid sending emails to addresses or domains on a blocklist or to only explicitly generate them
+   * for email address in an allow list.
+   */
+  email?: {
+    verificationRequest?: boolean
+  }
+  /** If Credentials provider is used, it contains the user credentials */
+  credentials?: Record<string, CredentialInput>
+}
+
+export interface RedirectCallbackParams {
+  /** URL provided as callback URL by the client */
+  url: string
+  /** Default base URL of site (can be used as fallback) */
+  baseUrl: string
+}
+
+export interface SessionCallbackParams<N = any> {
+  session: Session
+  /** Available when {@link AuthConfig.session} is set to `strategy: "jwt"` */
+  token?: JWT
+  /** Available when {@link AuthConfig.session} is set to `strategy: "database"`. */
+  user?: AdapterUser
+  /**
+   * Available when using {@link AuthConfig.session} `strategy: "database"` and an update is triggered for the session.
+   *
+   * :::note
+   * You should validate this data before using it.
+   * :::
+   */
+  newSession?: N
+  trigger?: "update"
+}
+
+export interface JwtCallbackParams<P = Profile, A = Account> {
+  /**
+   * When `trigger` is `"signIn"` or `"signUp"`, it will be a subset of {@link JWT},
+   * `name`, `email` and `image` will be included.
+   *
+   * Otherwise, it will be the full {@link JWT} for subsequent calls.
+   */
+  token: JWT
+  /**
+   * Either the result of the {@link OAuthConfig.profile} or the {@link CredentialsConfig.authorize} callback.
+   * @note available when `trigger` is `"signIn"` or `"signUp"`.
+   *
+   * Resources:
+   * - [Credentials Provider](https://authjs.dev/reference/core/providers_credentials)
+   * - [User database model](https://authjs.dev/reference/adapters#user)
+   */
+  user?: User | AdapterUser
+  /**
+   * Contains information about the provider that was used to sign in.
+   * Also includes {@link TokenSet}
+   * @note available when `trigger` is `"signIn"` or `"signUp"`
+   */
+  account?: A | null
```

Comment:
One drawback of making `user` and `account` optional is that we will have to check them for `undefined` even when `trigger` is `signIn` or `signUp`. This is a breaking change for existing TypeScript users.

Before this PR:

```typescript
callbacks: {
  async jwt({ token, user, account, trigger }) {
    if (trigger === "signIn") {
      console.log(user); // user is not User | AdapterUser
      console.log(account); // account is A | null
    }
  }
}
```

After this PR:

```typescript
callbacks: {
  async jwt({ token, user, account, trigger }) {
    if (trigger === "signIn") {
      console.log(user); // user is User | AdapterUser | undefined
      console.log(account); // account is A | null | undefined
    }
  }
}
```

Reply:
@Juneezee I use type union to help typescript developers [narrowing down](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions) the types. Unfortunately, this makes jsdocs not DRY anymore. Do you have any advice?